packages/components/src/internal/popup-utils.test.ts

```diff
@@ -0,0 +1,202 @@
+import { render } from '@testing-library/react';
+import { useRef } from 'react';
+import { div, h } from 'react-hyperscript-helpers';
+
+import { computePopupPosition, Position, Size, useBoundingRects } from './popup-utils';
+
+describe('computePopupPosition', () => {
+  it('computes position of popup element based on position of target element', () => {
+    // Arrange
+    const elementSize: Size = { width: 300, height: 100 };
+    const viewportSize: Size = {
+      width: 1280,
+      height: 960,
+    };
+    const targetPosition: Position = {
+      top: 430,
+      right: 690,
+      bottom: 530,
+      left: 590,
+    };
+    const gap = 10;
+
+    // Act
+    const topPopup = computePopupPosition({
+      elementSize,
+      gap,
+      preferredSide: 'top',
+      targetPosition,
+      viewportSize,
+    });
+
+    const rightPopup = computePopupPosition({
+      elementSize,
+      gap,
+      preferredSide: 'right',
+      targetPosition,
+      viewportSize,
+    });
+
+    const bottomPopup = computePopupPosition({
+      elementSize,
+      gap,
+      preferredSide: 'bottom',
+      targetPosition,
+      viewportSize,
+    });
+
+    const leftPopup = computePopupPosition({
+      elementSize,
+      gap,
+      preferredSide: 'left',
+      targetPosition,
+      viewportSize,
+    });
+
+    // Assert
+    expect(topPopup.position).toEqual({ top: 320, right: 790, bottom: 420, left: 490 });
+    expect(topPopup.side).toBe('top');
+
+    expect(rightPopup.position).toEqual({ top: 430, right: 1000, bottom: 530, left: 700 });
+    expect(rightPopup.side).toBe('right');
+
+    expect(bottomPopup.position).toEqual({ top: 540, right: 790, bottom: 640, left: 490 });
+    expect(bottomPopup.side).toBe('bottom');
+
+    expect(leftPopup.position).toEqual({ top: 430, right: 580, bottom: 530, left: 280 });
+    expect(leftPopup.side).toBe('left');
+  });
+
+  it('moves popup to opposite side if there is not enough space on preferred side', () => {
```

Comment:
Also add a test for top/bottom when there isn't enough room?

Reply:
Done in https://github.com/DataBiosphere/terra-ui/pull/4234/commits/148386274ecf087b324454e9924fc64f365492da.